server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -3159,7 +3169,11 @@ SnapshotsInProgress computeUpdatedState() {
                 }
                 final List<SnapshotsInProgress.Entry> newEntries = new ArrayList<>(oldEntries.size());
                 for (SnapshotsInProgress.Entry entry : oldEntries) {
-                    newEntries.add(applyToEntry(entry, updates.getValue()));
+                    final var newEntry = applyToEntry(entry, updates.getValue());
+                    newEntries.add(newEntry);
+                    if (newEntry != entry && newEntry.state().completed()) {
+                        newlyCompletedEntries.add(newEntry);
```

Comment:
I could use help to better understand two piece of information. 
1. How come a shard update for one snapshot end up updating the state of another snapshot?
2. How come a `ShardSnapshotUpdate` processed unassigned shards before `SnapshotsService#processExternalChanges`?

Using [this particular failure](https://github.com/elastic/elasticsearch/issues/104939#issuecomment-1918884740) as an example, I think the answer to the first question is the following sequence of calls when processing shard updates for `snapshot-clone-11`:

https://github.com/elastic/elasticsearch/blob/d1ca37a3fe29f93fecd6252f91390691601f8536/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L3236

https://github.com/elastic/elasticsearch/blob/d1ca37a3fe29f93fecd6252f91390691601f8536/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L3345

https://github.com/elastic/elasticsearch/blob/d1ca37a3fe29f93fecd6252f91390691601f8536/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L3380

https://github.com/elastic/elasticsearch/blob/d1ca37a3fe29f93fecd6252f91390691601f8536/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L3401

In `initShardSnapshotStatus`, we found out that the shard is unassigned and marked its state to `MISSING` which in turns leads to `snapshot-paritial-12` being marked as `SUCCESS`.

Is this correct?

Assuming the above is more or less correct, now for question 2, I don't understand how the unassigned shard was *not* processed `processExternalChanges` which is called inside `SnapshotsService#applyClusterState`. This is where my lack of knowledge for cluster coordination hurts ... I thought cluster state update and apply run strictly one after the other, i.e. after a cluster state update, the apply runs before computing another update. In regular case, node disconnection is handled by `processExternalChanges` which ends snapshots based on the missing shards. How come `ShardSnapshotUpdate` sees the missing shard first in this case? You said

> while that update was in flight the node left the cluster so the other shard snapshots moved to state MISSING in the same update

This seems to say while `ShardSnapshotUpdate` is running, the cluster state changed under it? Shard update uses a dedicated master task queue which I assume they cannot be processed together with node left? Sorry this all sounds pretty silly. I must have some serious/key misunderstanding somewhere.  I'd appreciate if you could help clarify the sequence/order of events here. Thanks!

Reply:
> In initShardSnapshotStatus, we found out that the shard is unassigned and marked its state to MISSING which in turns leads to snapshot-paritial-12 being marked as SUCCESS.

Correct.

> I don't understand how the unassigned shard was not processed processExternalChanges 

Because when `processExternalChanges` was called `[index-0][1]` was still in state `INIT` in `snapshot-clone-11`, and clones don't run on the data nodes so don't get cancelled on a node-left event, which means that the subsequent operations on `[index-0][1]` also remain unchanged (in state `QUEUED`). But then when that shard clone completes it moves to `SUCCESS` and the later `QUEUED` states all move to `MISSING` via the process you describe.

Reply:
We sync in a separate channel and it helped me to understand why `snapshot-partial-12` did not get finished during `processExternalChanges`. Because (1) the shard did not get marked as a `knownFailure` since it was part of a clone and (2) `snapshot-partial-12` at that time still has the shard as `UNASSIGNED_QUEUED` which does not has a nodeId and hence does not respond to node leaving.